### PR TITLE
[pom] Use ${revision} replace specific version.

### PR DESCRIPTION
**Pull request by joyCurry30**
As we use specific version like "3.0-SNAPSHOT", we need to repeat modifications between different modules. So why not use a placeholder like ${revision}.

### CODE REVIEW

**Review by whhe — COMMENTED** [review]
If 'revision' property is used, we should also add the `flatten-maven-plugin`, refer to https://maven.apache.org/maven-ci-friendly.html#install-deploy

**Comment by joyCurry30** [comment]
@whhe I add the flatten-maven-plugin. cc plz.

File: pom.xml [hunk 1]
```diff
@@ -465,6 +467,33 @@ under the License.
                     <doclint>none</doclint>
                 </configuration>
             </plugin>
+
+            <!-- make sure that flatten runs after shaded -->
```
Comment (lvyanquan) [inline pom.xml:471]: LGTM. minor suggestion: add comment to warn that this plugin takes effect for Maven version upper than 3.5.0.

**Review by lvyanquan — APPROVED** [review]

File: pom.xml (hunk 1 above)
Comment (whhe) [reply pom.xml:471]: +1

Comment (joyCurry30) [reply pom.xml:471]: Done.

File: pom.xml [hunk 2]
```diff
@@ -465,6 +467,34 @@ under the License.
...
+            <!-- This plugin takes effect for Maven version upper more than 3.5.0 -->
+            <plugin>
+                <groupId>org.codehaus.mojo</groupId>
+                <artifactId>flatten-maven-plugin</artifactId>
+                <version>${flatten-maven-plugin.version}</version>
+                <configuration>
+                    <updatePomFile>true</updatePomFile>
```
[the remark this answers is not in the document]
Comment (joyCurry30) [reply pom.xml:478]: Ok. I will add it.

[the remark this answers is not in the document]
Comment (joyCurry30) [reply pom.xml:478]: Ok. I will add it.

**Review by whhe — APPROVED** [review]
LGTM, wait for ci to pass.

**Comment by leonardBang** [comment]
Thanks @joyCurry30 and @whhe for the contribution and review, I've one minor question: What's the meaning `revision`? could we use maven default properties `${project.version}` in this case ?

**Comment by whhe** [comment]
> Thanks @joyCurry30 and @whhe for the contribution and review, I've one minor question: What's the meaning `revision`? could we use maven default properties `${project.version}` in this case ?

As we need to declare `${project.version}` explicitly in each pom file (the parent part), we need a new property to set a original value for it, and `revision` is a common option to be used here.

**Comment by leonardBang** [comment]
Thanks @whhe for the explanation, got it.

**Pull request merged**